command/recorderd/job_calendar.go

```diff
@@ -0,0 +1,515 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/bitmark-inc/logger"
+)
+
+const (
+	defaultNum                = 0
+	defaultTimeStrErrorMsg    = "clock time out of range"
+	defaultTimePeriodErrorMsg = "time period format error"
+	timePeriodSeparator       = ","
+	clockSeparator            = "-"
+	spaceChar                 = " "
+	allDayClockStr            = ""
+	defaultIndex              = 0
+	oneWeekDuration           = time.Duration(24*7) * time.Hour
+	delayOfStartStop          = time.Duration(5) * time.Second
+	JobCalendarPrefix         = "calendar"
+)
+
+type JobCalendar interface {
+	pickNextStartEvent(time.Time) interface{}
+	pickNextStopEvent(time.Time) interface{}
+	pickInitialiseStartEvent(time.Time) interface{}
+	pickInitialiseStopEvent(time.Time) interface{}
+	refresh(calendar ConfigCalendar)
+	rescheduleStartEventsPrior(event time.Time)
+	rescheduleStopEventsPrior(event time.Time)
+	setLog(l *logger.L)
+}
+
+type NumberRange struct {
+	min uint32
+	max uint32
+}
+
+// collapse events, all start time put in one place, all end time put in one place
+type FlattenEvents struct {
+	start []time.Time
+	stop  []time.Time
+}
+
+type SingleEvent struct {
+	start time.Time
+	stop  time.Time
+}
+
+type JobCalendarData struct {
+	flattenEvents     FlattenEvents
+	events            map[time.Weekday][]SingleEvent
+	rawData           ConfigCalendar
+	rescheduleChannel chan struct{}
+	log               *logger.L
+}
+
+type TimeData struct {
+	hour, minute uint32
+}
+
+func newJobCalendar(channel chan struct{}) JobCalendar {
```

Comment:
Is this channel only used for sending messages? Probably consider `chan<- struct{}`. Similar to the job manager.

Reply:
yeah, good idea